### DISCUSSION TITLE
How to effectively use crew::crew_controller_local?

### DISCUSSION THREAD

**Discussion by udurraniAtPresage** [hypebright/async_shiny#19] [Q&A] [answered]
**Goal**
In a two-tab shiny app, I render a ggiraph plot in the second tab which is not active by default. This is so that the plot renders while the user is still on the first tab. The code for the plot is in a separate module.
At the same time, there will be multiple simultaneous users of this app.

**Questions**
Is it wise to use `crew_controller_local(workers = 1, seconds_idle = 10)`? Each user creates their own session and the `crew_controller_local` expression is part of the plot module so they create their own separate R worker. Is that correct?

**Comment by hypebright** [comment, marked as answer]
Hi @udurraniAtPresage,

Yes, in Shiny, each user gets their own R session. So if the `crew_controller_local instance` is created as part of the server logic or module, each user session will create its own `crew_controller_local` object. Following that logic, each user creates its own separate R worker. This worker can be used for the plot module, if rendering that particular plot is a blocking operation.

This may sound like an open door perhaps, but you might also want to consider reducing the plot rendering time and move computations out of the plot elsewhere and send that to a worker. Most of the time plot rendering takes long because there's a lot of data behind it, but the amount of detail is not always necessary for the plot itself. Depends on your specific situation of course 😄.

Also, I don't know anything about the scale of your app, but be aware that for a high number of concurrent users, allocating a separate worker per user may strain server resources!

**Reply by udurraniAtPresage** [reply]
Thank you. Your explanation is very useful. I have allocated 16 GB memory and 4 CPUs for a single container instance where only 80 requests per instance are allowed. I expect 5-8 simultaneous users so I hope this would work fine with my setup.
